### PR TITLE
rustwide_builder: allow using a try build

### PR DESCRIPTION
**Pull request by jsha**
Followup from https://github.com/rust-lang/docs.rs/pull/1889#issuecomment-1302217471

### CODE REVIEW

**Comment by jsha** [comment]
I'm confused about the clippy errors, which seem unrelated to my change. Maybe there's a new version of clippy that triggers on some things it didn't before?

**Comment by GuillaumeGomez** [comment]
The new release was yesterday so new clippy lints available now. ;)

**Comment by Nemo157** [comment]
Ah, I was confused why my pre-fixing clippy lints as they appeared on nightly didn't catch these. I don't use `--all-targets` locally and these were in the test code, I'll remember to use it next time I do that (there's already some new warnings on the latest nightly clippy triggering).

File: src/docbuilder/rustwide_builder.rs [hunk 1]
```diff
@@ -108,6 +118,23 @@ impl RustwideBuilder {
...
 
     pub fn update_toolchain(&mut self) -> Result<bool> {
+        // For CI builds, a lot of the normal update_toolchain things don't apply.
+        // CI builds are only for one platform (https://forge.rust-lang.org/infra/docs/rustc-ci.html#try-builds)
+        // so we only try installing for the current platform. If that's not a match,
+        // for instance if we're running on macOS or Windows, this will error.
+        // Also, detecting the rustc version relies on calling rustc through rustup with the
+        // +channel argument, but the +channel argument doesn't work for CI builds. So
+        // we fake the rustc version and install from scratch every time since we can't detect
+        // the already-installed rustc version.
```
Comment (jyn514) [inline src/docbuilder/rustwide_builder.rs:128]: The + thing sounds like something we could fix upstream in rustwide.

**Review by jyn514 — APPROVED** [review]

**Comment by syphar** [comment]
@jyn514 is something missing to merge this?

**Comment by jyn514** [comment]
@syphar not that I'm aware of - haven't had the chance to test it locally, but hopefully @jsha has tested it. It would be nice to avoid the workaround by fixing rustwide but we shouldn't block on that, this is too useful haha

**Comment by syphar** [comment]
@jsha just in case: did you run tests with this code specifically?

**Comment by jsha** [comment]
Yes, I've run `cargo test` with `DOCSRS_TOOLCHAIN=6703d43029387eef125596dd7071d0d4fab902c3` in the env on this branch.

**Comment by syphar** [comment]
> Yes, I've run `cargo test` with `DOCSRS_TOOLCHAIN=6703d43029387eef125596dd7071d0d4fab902c3` in the env on this branch.
> 
> 

Cargo test won't test the build, can you check cargo test --ignored?

**Comment by syphar** [comment]
cc @jsha just as a last check, to be safe: did you run `cargo test --ignored` with the try-build? so the build-tests run?

**Comment by jsha** [comment]
I ran it after your suggestion (thanks!) and it failed, but I haven't had a chance to dig into why.

**Comment by syphar** [comment]
> I ran it after your suggestion (thanks!) and it failed, but I haven't had a chance to dig into why.

but the manual build with this code worked? 

Just a thought: (I don't know much about try-builds), but couldn't we just use some specific try-build that exists for longer, and add a testcase testing if running the build works with it?

**Comment by jsha** [comment]
> but the manual build with this code worked?

Yep.

> Just a thought: (I don't know much about try-builds), but couldn't we just use some specific try-build that exists for longer, and add a testcase testing if running the build works with it?

This is a good idea. I'll have to check how long they last.

Here's the error I'm getting from `cargo test -- --ignored`:

```
2022-11-13T09:01:54.990157Z  INFO log: [stderr] error: failed to create directory `/opt/rustwide/target/debug` log.target="rustwide::cmd" log.module_path="rustwide::cmd" log.file="/home/jsha/.cargo/registry/src/index.crates.io-e139d0d48fed7772/rustwide-0.15.1/src/cmd/mod.rs" log.line=621
2022-11-13T09:01:54.990222Z  INFO log: [stderr]  log.target="rustwide::cmd" log.module_path="rustwide::cmd" log.file="/home/jsha/.cargo/registry/src/index.crates.io-e139d0d48fed7772/rustwide-0.15.1/src/cmd/mod.rs" log.line=621
2022-11-13T09:01:54.990253Z  INFO log: [stderr] Caused by: log.target="rustwide::cmd" log.module_path="rustwide::cmd" log.file="/home/jsha/.cargo/registry/src/index.crates.io-e139d0d48fed7772/rustwide-0.15.1/src/cmd/mod.rs" log.line=621
2022-11-13T09:01:54.990279Z  INFO log: [stderr]   Permission denied (os error 13) log.target="rustwide::cmd" log.module_path="rustwide::cmd" log.file="/home/jsha/.cargo/registry/src/index.crates.io-e139d0d48fed7772/rustwide-0.15.1/src/cmd/mod.rs" log.line=621
```

**Comment by jyn514** [comment]
All commit artifacts are deleted every 6 months. If you want to update the try commit manually every 5 months that doesn't seem unreasonable, but given that this is just for local development I think it's ok to just not test it too.

**Comment by jsha** [comment]
This is interesting. If I run just one of the build tests at a time, they pass. For instance each of these passes:

```
cargo test -- --ignored test_build_binary_crate 
cargo test -- --ignored test_build_crate
```

But this fails:

    cargo test -- --ignored test_build_binary_crate test_build_crate

And this passes:

    cargo test -- --ignored test_build_binary_crate test_build_crate --test-threads 1

So it seems like some sort of concurrency issue, with different threads stomping on each other.

**Comment by syphar** [comment]
> All commit artifacts are deleted every 6 months. If you want to update the try commit manually every 5 months that doesn't seem unreasonable, but given that this is just for local development I think it's ok to just not test it too.

Ah, I didn't know that. 

Then yes, manual test is fine.

File: src/docbuilder/rustwide_builder.rs (hunk 1 above)
Comment (syphar) [reply src/docbuilder/rustwide_builder.rs:128]: This sounds like this is not blocking merging this PR, is it?

Comment (syphar) [reply src/docbuilder/rustwide_builder.rs:128]: ( For merging I need to resolve the conversation :) )

Comment (jyn514) [reply src/docbuilder/rustwide_builder.rs:128]: No, not a.blocker. would be nice to file an issue before it gets resolved and we lose the comment forever though

Comment (syphar) [reply src/docbuilder/rustwide_builder.rs:128]: kk, I don't know enought about this to create the issue, so I'm waiting for one of you two to do  it, and then also merge this PR.

**Comment by jsha** [comment]
> So it seems like some sort of concurrency issue, with different threads stomping on each other.

Scratch that. I realized I was actually running into this concurrency issue with `DOCRS_TOOLCHAIN=nightly`, which suggests it's a separate, preexisting issue.

Once I switched back to `DOCSRS_TOOLCHAIN=6703d43029387eef125596dd7071d0d4fab902c3` I get this error (which I have seen before when I was correctly setting DOCSRS_TOOLCHAIN):

    cargo test -- --ignored test_build_crate
    ...
    thread 'docbuilder::rustwide_builder::tests::test_build_crate' panicked at 'called `Result::unwrap()` on an `Err` value: failed to remove '.workspace/builds' : Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', src/docbuilder/rustwide_builder.rs:876:58

_Also_, I realized the rustwide build tests take their environment from the TestEnvironment, so I would expect them to actually ignore environment variables. Gonna need to do a little more digging.

**Comment by syphar** [comment]
> > So it seems like some sort of concurrency issue, with different threads stomping on each other.
> 
> Scratch that. I realized I was actually running into this concurrency issue with `DOCRS_TOOLCHAIN=nightly`, which suggests it's a separate, preexisting issue.

just fyi: in CI we're running the build-tests also with `test-threads=1`: 
https://github.com/rust-lang/docs.rs/blob/161dba0b659b133647f880adc81c0a940bf9a1a6/.github/workflows/ci.yml#L118

**Comment by jsha** [comment]
> I realized the rustwide build tests take their environment from the TestEnvironment, so I would expect them to actually ignore environment variables.

Figured that out. The test env does call from_env, and then overrides a handful of values:

https://github.com/rust-lang/docs.rs/blob/161dba0b659b133647f880adc81c0a940bf9a1a6/src/test/mod.rs#L258-L283

We should probably make TestEnvironment more hermetic by setting all (or almost all?) of the fields, but for now I've let it be.

With the latest pushed commit, this now passes the crate build tests with `DOCSRS_TOOLCHAIN=6703d43029387eef125596dd7071d0d4fab902c3`.

File: src/docbuilder/rustwide_builder.rs (hunk 1 above)
Comment (jsha) [reply src/docbuilder/rustwide_builder.rs:128]: https://github.com/rust-lang/rustwide/issues/75

**Pull request merged**